src/oauth-provider.ts

```diff
@@ -1010,6 +1048,55 @@ class OAuthProviderImpl {
     }
   }
 
+  /**
+   * Decodes a token and returns token data with decrypted props
+   * @param token - The granted token
+   * @param env - Cloudflare Worker environment variables
+   * @returns Promise resolving to token data with decrypted props, or null if token is invalid
+   */
+  async unwrapToken<T = any>(token: string, env: any): Promise<TokenSummary<T> | null> {
+    const parts = token.split(':');
+    const isPossiblyInternalFormat = parts.length === 3;
+
+    if (!isPossiblyInternalFormat) {
+      return null;
+    }
+
+    // Retrieve the token from KV
+    const [userId, grantId] = parts;
+    const id = await generateTokenId(token);
+    const tokenData: Token | null = await env.OAUTH_KV.get(`token:${userId}:${grantId}:${id}`, { type: 'json' });
+
+    // Return null if missing or expired
+    if (!tokenData) {
+      return null;
+    }
+    const now = Math.floor(Date.now() / 1e3);
+    if (tokenData.expiresAt < now) {
+      return null;
+    }
+
+    // Decrypt the props
+    const encryptionKey = await unwrapKeyWithToken(token, tokenData.wrappedEncryptionKey);
+    const decryptedProps = await decryptProps(encryptionKey, tokenData.grant.encryptedProps);
+
+    // Return the token data with decrypted instead of encrypted props
+    const { grant } = tokenData;
+    return {
+      id: tokenData.id,
+      grantId: tokenData.grantId,
+      userId: tokenData.userId,
+      createdAt: tokenData.createdAt,
+      expiresAt: tokenData.expiresAt,
+      audience: tokenData.audience,
+      grant: {
+        clientId: grant.clientId,
+        scope: grant.scope,
+        props: decryptedProps as T,
+      },
+    };
```

Comment:
Should this not include an `active` property as required by the OAuth 2.0 Token Introspection spec?

https://datatracker.ietf.org/doc/html/rfc7662#section-2.2

More generally, this response doesn't appear to be spec-compliant. Several required or recommended fields are either missing (active) or represented differently (exp, iat, sub, etc.), and some fields here don’t map cleanly to the introspection response at all.

Is this endpoint intentionally not an RFC 7662 introspection response (i.e. an internal representation), or is partial compliance acceptable here? If it’s intended to be an introspection endpoint, I think we should either align more closely with the spec or document the deviations explicitly.

Reply:
Yeah I didn't add the endpoint, but the `unwrapToken` method can sit behind a custom route and do the same. Main goal was to surface a method that could read tokens without having to look through the code and copy the "generateTokenId" machinery.
Can implement a proper RFC 7662 endpoint in another PR too.